### PR TITLE
Mutual authentication Implementation

### PR DESCRIPTION
**Pull request by ihsandemir**
Added mutual authentication capability to the C++ client.
    
Also a few fixes on the peer verification mode of the ssl socket.

For windows, we had to define an environment variable `SSL_CERT_FILE=C:\cacert.pem` and downloaded the file from https://curl.se/docs/caextract.html which is Mozilla CA store. This way OpenSSL library can find the intermediate CA authorities correctly and the test `ssl_enabled_trust_default_certificates` works as expected.

### CODE REVIEW

**Review by sancar — APPROVED** [review]

File: scripts/start-rc.sh [hunk 1]
```diff
@@ -65,7 +65,17 @@ else
         exit 1
     fi
 fi
-CLASSPATH="hazelcast-remote-controller-${HAZELCAST_RC_VERSION}.jar:hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}.jar:hazelcast-${HAZELCAST_TEST_VERSION}-tests.jar"
+if [ -f "hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}-tests.jar" ]; then
+echo "hazelcast-enterprise-tests.jar already exists, not downloading from maven."
+else
+    echo "Downloading: hazelcast enterprise jar com.hazelcast:hazelcast-enterprise:${HAZELCAST_ENTERPRISE_VERSION}:jar:tests"
```
Comment (yemreinci) [inline scripts/start-rc.sh:71]:
```suggestion
    echo "Downloading: hazelcast enterprise test jar com.hazelcast:hazelcast-enterprise:${HAZELCAST_ENTERPRISE_VERSION}:jar:tests"
```

File: scripts/start-rc.sh [hunk 2]
```diff
@@ -65,7 +65,17 @@ else
...
+    mvn -q dependency:get -DrepoUrl=${ENTERPRISE_REPO} -Dartifact=com.hazelcast:hazelcast-enterprise:${HAZELCAST_ENTERPRISE_VERSION}:jar:tests -Ddest=hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}-tests.jar
+    if [ $? -ne 0 ]; then
+        echo "Failed download hazelcast enterprise test jar com.hazelcast:hazelcast-enterprise:${HAZELCAST_ENTERPRISE_VERSION}:jar:tests"
+        exit 1
+    fi
+fi
+CLASSPATH="hazelcast-remote-controller-${HAZELCAST_RC_VERSION}.jar:hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}.jar:hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}-tests.jar:hazelcast-${HAZELCAST_TEST_VERSION}-tests.jar"
```
Comment (yemreinci) [inline scripts/start-rc.sh:78]:
```suggestion
CLASSPATH="\
hazelcast-remote-controller-${HAZELCAST_RC_VERSION}.jar:\
hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}.jar:\
hazelcast-enterprise-${HAZELCAST_ENTERPRISE_VERSION}-tests.jar:\
hazelcast-${HAZELCAST_TEST_VERSION}-tests.jar"
```

File: scripts/start-rc.bat [hunk 3]
```diff
@@ -50,7 +50,17 @@ if exist "hazelcast-enterprise-%HAZELCAST_ENTERPRISE_VERSION%.jar" (
     )
 )
 
-set CLASSPATH="hazelcast-remote-controller-%HAZELCAST_RC_VERSION%.jar;hazelcast-enterprise-%HAZELCAST_ENTERPRISE_VERSION%.jar;hazelcast-%HAZELCAST_TEST_VERSION%-tests.jar"
+if exist "hazelcast-enterprise-%HAZELCAST_ENTERPRISE_VERSION%-tests.jar" (
+    echo "hazelcast-enterprise-%HAZELCAST_ENTERPRISE_VERSION%-tests.jar already exists, not downloading from maven."
+) else (
+    echo "Downloading: hazelcast enterprise jar com.hazelcast:hazelcast-enterprise:%HAZELCAST_ENTERPRISE_VERSION%:jar:tests"
```
Comment (yemreinci) [inline scripts/start-rc.bat:56]:
```suggestion
    echo "Downloading: hazelcast enterprise test jar com.hazelcast:hazelcast-enterprise:%HAZELCAST_ENTERPRISE_VERSION%:jar:tests"
```

File: scripts/start-rc.bat [hunk 4]
```diff
@@ -50,7 +50,17 @@ if exist "hazelcast-enterprise-%HAZELCAST_ENTERPRISE_VERSION%.jar" (
...
+    call mvn -q dependency:get -DrepoUrl=%ENTERPRISE_REPO% -Dartifact=com.hazelcast:hazelcast-enterprise:%HAZELCAST_ENTERPRISE_VERSION%:jar:tests -Ddest=hazelcast-enterprise-%HAZELCAST_ENTERPRISE_VERSION%-tests.jar || (
+        echo "Failed download hazelcast enterprise jar com.hazelcast:hazelcast-enterprise:%HAZELCAST_ENTERPRISE_VERSION%:jar:tests"
```
Comment (yemreinci) [inline scripts/start-rc.bat:58]:
```suggestion
        echo "Failed download hazelcast enterprise test jar com.hazelcast:hazelcast-enterprise:%HAZELCAST_ENTERPRISE_VERSION%:jar:tests"
```

File: hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h [hunk 5]
```diff
@@ -56,8 +56,8 @@ namespace hazelcast {
                     spi::ClientContext &client_context_;
                     boost::asio::io_context &io_;
                     boost::asio::ip::tcp::resolver &io_resolver_;
+                    std::shared_ptr<boost::asio::ssl::context> ssl_context_;
```
Comment (yemreinci) [inline hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h:59]: Does this compile without OpenSSL or `boost/asio/ssl/context.hpp`?

**Review by yemreinci — CHANGES_REQUESTED** [review]

**Comment by yemreinci** [comment]
Doesn't compile without SSL:

```
$ cmake -DWITH_OPENSSL=0
$ make
...
/usr/local/include/boost/asio/ssl/detail/openssl_types.hpp:23:10: fatal error: 'openssl/conf.h' file not found
...
```

**Comment by ihsandemir** [comment]
verify-windows

**Comment by ihsandemir** [comment]
verify-windows

File: hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h [hunk 6]
```diff
@@ -56,8 +56,8 @@ namespace hazelcast {
...
 #ifdef HZ_BUILD_WITH_SSL
```
Comment (mdumandag) [inline hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h:60]: This 
```
#ifdef HZ_BUILD_WITH_SSL
#endif
```
do nothing now

File: hazelcast/src/hazelcast/client/network.cpp [hunk 7]
```diff
@@ -1287,6 +1293,8 @@ namespace hazelcast {
                 }
 
                 void SSLSocket::post_connect() {
+                    socket_.set_verify_mode(
+                            boost::asio::ssl::verify_peer | boost::asio::ssl::verify_fail_if_no_peer_cert);
```
Comment (mdumandag) [inline hazelcast/src/hazelcast/client/network.cpp:1297]: what if the user explicitly sets the verify mode in the SSL context via https://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/ssl__context/set_verify_mode.html? 

I think we should respect it

Comment (yemreinci) [reply hazelcast/src/hazelcast/client/network.cpp:1297]: Also, it says [here](https://www.openssl.org/docs/man1.0.2/man3/SSL_CTX_set_verify.html) that `SSL_VERIFY_FAIL_IF_NO_PEER_CERT` is ignored on the client side. Does it have any use here?

Comment (ihsandemir) [reply hazelcast/src/hazelcast/client/network.cpp:1297]: Good point. Changed it to respect user config. 
(I initially thought that it was a good idea to enforce but after a second thought I changed my mind on that.)

**Comment by ihsandemir** [comment]
> Doesn't compile without SSL:
> 
> ```
> $ cmake .. -DWITH_OPENSSL=0
> $ make
> ...
> /usr/local/include/boost/asio/ssl/detail/openssl_types.hpp:23:10: fatal error: 'openssl/conf.h' file not found
> ...
> ```

Good catch. Changed and tested it.

File: hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h (hunk 5 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h]: fixed.

File: hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h (hunk 6 above)
Comment (ihsandemir) [reply hazelcast/include/hazelcast/client/internal/socket/SocketFactory.h:59]: fixed.

**Review by mdumandag — APPROVED** [review]

File: examples/tls/BasicTLSClient.cpp [hunk 8]
```diff
@@ -23,9 +23,12 @@ int main() {
     hazelcast::client::address serverAddress("127.0.0.1", 5701);
     config.get_network_config().add_address(serverAddress);
 
+    boost::asio::ssl::context ctx(boost::asio::ssl::context::method::tlsv12_client);
+    ctx.set_default_verify_paths();
+    ctx.load_verify_file("/path/to/my/server/public/certificate");
```
Comment (yemreinci) [inline examples/tls/BasicTLSClient.cpp:28]: Should we also do `ctx.set_verify_mode(ssl::verify_peer)` to make sure the certificate is actually verified? What is the default behavior if you don't set the verify mode?

**Review by sancar — APPROVED** [review]

File: examples/tls/BasicTLSClient.cpp (hunk 8 above)
Comment (ihsandemir) [reply examples/tls/BasicTLSClient.cpp:28]: Added.

**Review by sancar — APPROVED** [review]

File: hazelcast/test/src/HazelcastTests1.cpp [hunk 9]
```diff
@@ -864,19 +1155,12 @@ namespace hazelcast {
...
                         std::chrono::seconds(2)).set_initial_backoff_duration(std::chrono::milliseconds(100));
-                config.set_cluster_name(get_ssl_cluster_name()).get_network_config().add_address(
-                        address("8.8.8.8", 5701)).get_ssl_config().set_enabled(true).add_verify_file(
-                        get_ca_file_path());
-                ASSERT_THROW(hazelcast_client client{new_client(std::move(config)).get()},
-                             exception::illegal_state);
-            }
+                boost::asio::ssl::context ctx(boost::asio::ssl::context::method::tlsv12_client);
+                ctx.set_default_verify_paths();
+                ctx.load_verify_file(get_ca_file_path());
```
Comment (yemreinci) [inline hazelcast/test/src/HazelcastTests1.cpp:1160]: I believe these are not relevant to this test. They are redundant without setting the verify mode anyways.

```suggestion
```

Comment (ihsandemir) [reply hazelcast/test/src/HazelcastTests1.cpp:1160]: removed.

**Review by yemreinci — APPROVED** [review]

**Pull request merged**